go/vt/vtgate/healthcheck.go

```diff
@@ -0,0 +1,279 @@
+		connTimeout: connTimeout,
+		retryDelay:  retryDelay,
+	}
+	hc.initConns(endPoints)
+	go hc.onAddrUpdate(addrUpdate)
+	return hc
+}
+
+// HealthCheck performs health checking and notifies downstream components about any changes.
+type HealthCheck struct {
+	mu          sync.RWMutex
+	addrToConns map[string]*healthCheckConn                              // addrToConns maps from address to the healthCheckConn object.
+	targetToEPs map[string]map[string]map[pbt.TabletType][]*pbt.EndPoint // targetToEPs maps from keyspace/shard/tablettype to a list of endpoints.
+	connTimeout time.Duration
+	retryDelay  time.Duration
+}
+
+// healthCheckConn contains details about an endpoint.
+type healthCheckConn struct {
+	mu         sync.RWMutex
+	target     *pbq.Target
+	cancelFunc context.CancelFunc
+	conn       tabletconn.TabletConn
+	stats      *pbq.RealtimeStats
+}
+
+// initConns starts health checking for the initial set of endpoints.
+// It waits till all endpoints are tried to connect once.
+func (hc *HealthCheck) initConns(endPoints []*pbt.EndPoint) {
+	var wg sync.WaitGroup
+	for _, endPoint := range endPoints {
+		wg.Add(1)
+		go hc.checkConn(endPoint, &wg)
+	}
+	wg.Wait()
+}
+
+// checkConn performs health checking on the given endpoint.
+func (hc *HealthCheck) checkConn(endPoint *pbt.EndPoint, wg *sync.WaitGroup) {
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	hcc := &healthCheckConn{
+		cancelFunc: cancelFunc,
+	}
+	addr := hc.endPointToString(endPoint)
+
+	defer func() {
+		// Clean up the healthCheckConn from maps, and notify addrwatcher
+		hc.mu.Lock()
+		// TODO: notify downstream component for endpoint going down
+		if hcc.target != nil {
+			hc.deleteEndPointFromTarget(hcc.target, endPoint)
+			delete(hc.addrToConns, addr)
+		}
+		hc.mu.Unlock()
+	}()
+
+	var conn tabletconn.TabletConn
+	var stream <-chan *pbq.StreamHealthResponse
+	var errfunc tabletconn.ErrFunc
+	var err error
+
+	// func to read one health check response, and notify downstream component
+	processResponse := func() error {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case shr, ok := <-stream:
+			if !ok {
+				return errfunc()
+			}
+			if shr.Target == nil || shr.RealtimeStats == nil || shr.RealtimeStats.HealthError != "" {
+				return fmt.Errorf("health stats is not valid: %v", shr)
+			}
+
+			if hcc.target == nil {
+				// The first time we see response for the endpoint.
+				// Add the healthCheckConn to maps, and notify downstream component for endpoint going up.
+				hc.mu.Lock()
+				hc.addrToConns[addr] = hcc
+				hc.addEndPointToTarget(shr.Target, endPoint)
+				// TODO: notify downstream component for endpoint going up
+				hc.mu.Unlock()
+			} else if hcc.target.TabletType != shr.Target.TabletType {
+				// Update targetToConn when tablettype changes.
+				hc.mu.Lock()
+				hc.deleteEndPointFromTarget(hcc.target, endPoint)
+				hc.addEndPointToTarget(shr.Target, endPoint)
+				hc.mu.Unlock()
+			}
+			hcc.mu.Lock()
```

Comment:
I changed the map targetToConn to exclude tablet type, as currently we do not have a need to distinguish among tablet types in the health check module. It also simplifies the logic here.